File: src/main/java/org/moinex/ui/common/CreditCardPaneController.java
/*
 * Filename: CreditCardPaneController.java
 * Created on: October 20, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.common;

import com.jfoenix.controls.JFXButton;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.YearMonth;
import java.time.format.DateTimeFormatter;
import java.util.List;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressBar;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.VBox;
import lombok.NoArgsConstructor;
import org.moinex.entities.CreditCard;
import org.moinex.services.CreditCardService;
import org.moinex.ui.dialog.AddCreditCardDebtController;
import org.moinex.ui.dialog.CreditCardInvoicePaymentController;
import org.moinex.ui.dialog.EditCreditCardController;
import org.moinex.ui.main.CreditCardController;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Credit Card Pane
 *
 * @note prototype is necessary so that each scene knows to which credit card it is
 * associated
 */
@Controller
@Scope("prototype") // Each instance of this controller is unique
@NoArgsConstructor
public class CreditCardPaneController
{
    @FXML
    private VBox rootVBox;

    @FXML
    private ImageView crcOperatorIcon;

    @FXML
    private Label crcName;

    @FXML
    private Label crcOperator;

    @FXML
    private Label limitLabel;

    @FXML
    private Label pendingPaymentsLabel;

    @FXML
    private Label limitAvailableLabel;

    @FXML
    private Label closureDayLabel;

    @FXML
    private Label nextInvoiceLabel;

    @FXML
    private Label dueDateLabel;

    @FXML
    private Label invoiceStatus;

    @FXML
    private Label invoiceTotal;

    @FXML
    private Label invoiceMonth;

    @FXML
    private Label limitProgressLabel;

    @FXML
    private JFXButton prevButton;

    @FXML
    private JFXButton nextButton;

    @FXML
    private ProgressBar limitProgressBar;

    @Autowired
    private ConfigurableApplicationContext springContext;

    @Autowired
    private CreditCardController creditCardController;

    private YearMonth currentDisplayedMonth;

    private CreditCardService creditCardService;

    private CreditCard creditCard;

    /**
     * Constructor
     * @param creditCardService CreditCardService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public CreditCardPaneController(CreditCardService creditCardService)
    {
        this.creditCardService = creditCardService;
    }

    @FXML
    private void initialize()
    {
        currentDisplayedMonth = YearMonth.now();
    }

    @FXML
    private void handleAddDebt()
    {
        WindowUtils.openModalWindow(
            Constants.ADD_CREDIT_CARD_DEBT_FXML,
            "Add Credit Card Debt",
            springContext,
            (AddCreditCardDebtController controller)
                -> { controller.setCreditCard(creditCard); },
            List.of(() -> creditCardController.updateDisplay()));
    }

    @FXML
    private void handleEditCreditCard()
    {
        WindowUtils.openModalWindow(
            Constants.EDIT_CREDIT_CARD_FXML,
            "Edit Credit Card",
            springContext,
            (EditCreditCardController controller)
                -> { controller.setCreditCard(creditCard); },
            List.of(() -> creditCardController.updateDisplay()));
    }

    @FXML
    private void handleArchiveCreditCard()
    {
        if (WindowUtils.showConfirmationDialog(
                "Confirmation",
                "Archive credit card " + creditCard.getName(),
                "Are you sure you want to archive this credit card?"))
        {
            try
            {
                creditCardService.archiveCreditCard(creditCard.getId());

                WindowUtils.showSuccessDialog("Success",
                                              "Credit card archived",
                                              "Credit card " + creditCard.getName() +
                                                  " has been archived");

                // Update credit card display in the main window
                creditCardController.updateDisplay();
            }
            catch (RuntimeException e)
            {
                WindowUtils.showErrorDialog("Error",
                                            "Error archiving credit card",
                                            e.getMessage());
                return;
            }
        }
    }

    @FXML
    private void handleDeleteCreditCard()
    {
        // Prevent the removal of a credit card with associated debts
        if (creditCardService.getDebtCountByCreditCard(creditCard.getId()) > 0)
        {
            WindowUtils.showErrorDialog(
                "Error",
                "Credit card has debts",
                "Cannot delete a credit card with associated debts");
            return;
        }

        if (WindowUtils.showConfirmationDialog(
                "Confirmation",
                "Delete credit card " + creditCard.getName(),
                "Are you sure you want to remove this credit card?"))
        {
            try
            {
                creditCardService.deleteCreditCard(creditCard.getId());

                WindowUtils.showSuccessDialog("Success",
                                              "Credit card deleted",
                                              "Credit card " + creditCard.getName() +
                                                  " has been deleted");

                // Update credit card display in the main window
                creditCardController.updateDisplay();
            }
            catch (RuntimeException e)
            {
                WindowUtils.showErrorDialog("Error",
                                            "Error removing credit card",
                                            e.getMessage());
                return;
            }
        }
    }

    @FXML
    private void handlePrevMonth()
    {
        currentDisplayedMonth = currentDisplayedMonth.minusMonths(1);
        updateInvoiceInfo();
    }

    @FXML
    private void handleNextMonth()
    {
        currentDisplayedMonth = currentDisplayedMonth.plusMonths(1);
        updateInvoiceInfo();
    }

    @FXML
    private void handleRegisterPayment()
    {
        WindowUtils.openModalWindow(
            Constants.CREDIT_CARD_INVOICE_PAYMENT_FXML,
            "Register Payment",
            springContext,
            (CreditCardInvoicePaymentController controller)
                -> { controller.setCreditCard(creditCard, currentDisplayedMonth); },
            // Update the display after the payment is registered with the current month
            List.of(
                () -> { creditCardController.updateDisplay(currentDisplayedMonth); }));
    }

    /**
     * Load the Credit Card Pane
     * @param creditCard Credit Card to load
     * @return The updated VBox
     */
    public VBox updateCreditCardPane(CreditCard crc, YearMonth month)
    {
        // If the crc is null, do not update the pane
        if (crc == null)
        {
            setDefaultValues();
            return rootVBox;
        }

        this.creditCard            = crc;
        this.currentDisplayedMonth = month;

        crcName.setText(creditCard.getName());
        crcOperator.setText(creditCard.getOperator().getName());
        crcOperatorIcon.setImage(new Image(Constants.CRC_OPERATOR_ICONS_PATH +
                                           creditCard.getOperator().getIcon()));

        BigDecimal limit = creditCard.getMaxDebt();
        BigDecimal pendingPayments =
            creditCardService.getTotalPendingPayments(creditCard.getId());

        BigDecimal limitAvailable =
            creditCardService.getAvailableCredit(creditCard.getId());

        limitLabel.setText(UIUtils.formatCurrency(limit));

        pendingPaymentsLabel.setText(UIUtils.formatCurrency(pendingPayments));

        limitAvailableLabel.setText(UIUtils.formatCurrency(limitAvailable));

        // Set percentage of the usage of the limit
        BigDecimal limitProgress =
            limit.compareTo(BigDecimal.ZERO) == 0
                ? BigDecimal.ZERO
                : pendingPayments.divide(limit, 2, RoundingMode.HALF_UP);

        limitProgressBar.setProgress(limitProgress.doubleValue());
        limitProgressLabel.setText(
            UIUtils.formatPercentage(limitProgress.doubleValue() * 100));

        dueDateLabel.setText(creditCard.getBillingDueDay().toString());

        closureDayLabel.setText(creditCard.getClosingDay().toString());

        // Fromat LocalDateTime to MM/YYYY
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM/yy");

        nextInvoiceLabel.setText(
            creditCardService.getNextInvoiceDate(creditCard.getId()).format(formatter));

        updateInvoiceInfo();

        return rootVBox;
    }

    /**
     * Update the invoice information
     */
    public void updateInvoiceInfo()
    {
        if (creditCard == null)
        {
            return;
        }

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM/yy");

        invoiceMonth.setText(currentDisplayedMonth.format(formatter));

        invoiceTotal.setText(UIUtils.formatCurrency(
            creditCardService.getInvoiceAmount(creditCard.getId(),
                                               currentDisplayedMonth.getMonthValue(),
                                               currentDisplayedMonth.getYear())));

        invoiceStatus.setText(
            creditCardService
                .getInvoiceStatus(creditCard.getId(),
                                  currentDisplayedMonth.getMonthValue(),
                                  currentDisplayedMonth.getYear())
                .toString());
    }

    /**
     * Set the default values for the credit card pane
     */
    private void setDefaultValues()
    {
        crcName.setText("");
        crcOperator.setText("");
        limitLabel.setText("");
        pendingPaymentsLabel.setText("");
        limitAvailableLabel.setText("");
        closureDayLabel.setText("");
        nextInvoiceLabel.setText("");
        dueDateLabel.setText("");
        invoiceStatus.setText("");
        invoiceTotal.setText("");
        invoiceMonth.setText("");
        crcOperatorIcon.setImage(new Image(Constants.DEFAULT_ICON));
    }
}

File: src/main/kotlin/org/moinex/ui/common/CreditCardPaneController.kt
/*
 * Filename: CreditCardPaneController.kt (original filename: CreditCardPaneController.java)
 * Created on: October 20, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 19/03/2026
 */

package org.moinex.ui.common

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.control.ProgressBar
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.layout.VBox
import org.moinex.common.constant.Files
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.extension.isZero
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.creditcard.CreditCard
import org.moinex.service.PreferencesService
import org.moinex.service.creditcard.CreditCardService
import org.moinex.ui.dialog.creditcard.AddCreditCardCreditController
import org.moinex.ui.dialog.creditcard.AddCreditCardDebtController
import org.moinex.ui.dialog.creditcard.CreditCardCreditsController
import org.moinex.ui.dialog.creditcard.CreditCardInvoicePaymentController
import org.moinex.ui.dialog.creditcard.EditCreditCardController
import org.moinex.ui.main.CreditCardController
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.MessageFormat
import java.time.YearMonth

@Controller
@Scope("prototype")
class CreditCardPaneController(
    private val creditCardService: CreditCardService,
    private val springContext: ConfigurableApplicationContext,
    private val creditCardController: CreditCardController,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var rootVBox: VBox

    @FXML
    private lateinit var crcOperatorIcon: ImageView

    @FXML
    private lateinit var crcName: Label

    @FXML
    private lateinit var crcOperator: Label

    @FXML
    private lateinit var limitLabel: Label

    @FXML
    private lateinit var pendingPaymentsLabel: Label

    @FXML
    private lateinit var availableLimitLabel: Label

    @FXML
    private lateinit var availableRebateLabel: Label

    @FXML
    private lateinit var closureDayLabel: Label

    @FXML
    private lateinit var nextInvoiceLabel: Label

    @FXML
    private lateinit var dueDateLabel: Label

    @FXML
    private lateinit var invoiceStatusLabel: Label

    @FXML
    private lateinit var invoiceMonthLabel: Label

    @FXML
    private lateinit var limitProgressLabel: Label

    @FXML
    private lateinit var invoiceMonthNavigatorBarLabel: Label

    @FXML
    private lateinit var limitProgressBar: ProgressBar

    private var currentDisplayedMonth = YearMonth.now()
    private lateinit var creditCard: CreditCard

    companion object {
        private const val GREEN_LABEL_STYLE = "-fx-text-fill: green"
        private const val BLACK_LABEL_STYLE = "-fx-text-fill: black"
    }

    @FXML
    private fun initialize() {
        currentDisplayedMonth = YearMonth.now()
    }

    @FXML
    private fun handleAddDebt() {
        WindowUtils.openModalWindow(
            Files.ADD_CREDIT_CARD_DEBT_FXML,
            preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_MODAL_ADD_DEBT),
            springContext,
            { controller: AddCreditCardDebtController -> controller.setCreditCard(creditCard) },
            listOf(Runnable { creditCardController.updateDisplay() }),
        )
    }

    @FXML
    private fun handleAddCredit() {
        WindowUtils.openModalWindow(
            Files.ADD_CREDIT_CARD_CREDIT_FXML,
            preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_MODAL_ADD_CREDIT),
            springContext,
            { controller: AddCreditCardCreditController -> controller.setCreditCard(creditCard) },
            listOf(Runnable { creditCardController.updateDisplay() }),
        )
    }

    @FXML
    private fun handleEditCreditCard() {
        WindowUtils.openModalWindow(
            Files.EDIT_CREDIT_CARD_FXML,
            preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_MODAL_EDIT),
            springContext,
            { controller: EditCreditCardController -> controller.setCreditCard(creditCard) },
            listOf(Runnable { creditCardController.updateDisplay() }),
        )
    }

    @FXML
    private fun handleArchiveCreditCard() {
        if (WindowUtils.showConfirmationDialog(
                MessageFormat.format(
                    preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_ARCHIVE_TITLE),
                    creditCard.name,
                ),
                preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_ARCHIVE_MESSAGE),
                preferencesService.bundle,
            )
        ) {
            runCatching {
                creditCardService.archiveCreditCard(creditCard.id!!)

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_ARCHIVE_SUCCESS_TITLE),
                    MessageFormat.format(
                        preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_ARCHIVE_SUCCESS_MESSAGE),
                        creditCard.name,
                    ),
                )

                creditCardController.updateDisplay()
            }.onFailure { e ->
                when (e) {
                    is EntityNotFoundException, is IllegalStateException ->
                        WindowUtils.showErrorDialog(
                            preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_ARCHIVE_ERROR),
                            e.message ?: "",
                        )
                    else -> throw e
                }
            }
        }
    }

    @FXML
    private fun handleDeleteCreditCard() {
        if (creditCardService.getDebtCountByCreditCard(creditCard.id!!) > 0) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_DELETE_HAS_DEBTS_TITLE),
                preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_DELETE_HAS_DEBTS_MESSAGE),
            )
            return
        }

        if (WindowUtils.showConfirmationDialog(
                MessageFormat.format(
                    preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_DELETE_TITLE),
                    creditCard.name,
                ),
                preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_DELETE_MESSAGE),
                preferencesService.bundle,
            )
        ) {
            runCatching {
                creditCardService.deleteCreditCard(creditCard.id!!)

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_DELETE_SUCCESS_TITLE),
                    MessageFormat.format(
                        preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_DELETE_SUCCESS_MESSAGE),
                        creditCard.name,
                    ),
                )

                creditCardController.updateDisplay()
            }.onFailure { e ->
                when (e) {
                    is EntityNotFoundException, is IllegalStateException ->
                        WindowUtils.showErrorDialog(
                            preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_DIALOG_DELETE_ERROR),
                            e.message ?: "",
                        )
                    else -> throw e
                }
            }
        }
    }

    @FXML
    private fun handleShowRebates() {
        WindowUtils.openModalWindow(
            Files.CREDIT_CARD_CREDITS_FXML,
            preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_MODAL_SHOW_CREDITS),
            springContext,
            { _: CreditCardCreditsController -> },
            listOf(Runnable { creditCardController.updateDisplay() }),
        )
    }

    @FXML
    private fun handlePrevMonth() {
        currentDisplayedMonth = currentDisplayedMonth.minusMonths(1)
        updateInvoiceInfo()
    }

    @FXML
    private fun handleNextMonth() {
        currentDisplayedMonth = currentDisplayedMonth.plusMonths(1)
        updateInvoiceInfo()
    }

    @FXML
    private fun handleRegisterPayment() {
        WindowUtils.openModalWindow(
            Files.CREDIT_CARD_INVOICE_PAYMENT_FXML,
            preferencesService.translate(TranslationKeys.COMMON_CREDIT_CARD_MODAL_REGISTER_PAYMENT),
            springContext,
            { controller: CreditCardInvoicePaymentController ->
                controller.setCreditCard(creditCard, currentDisplayedMonth)
            },
            listOf(Runnable { creditCardController.updateDisplay(currentDisplayedMonth) }),
        )
    }

    fun updateCreditCardPane(
        crc: CreditCard,
        month: YearMonth,
    ): VBox {
        creditCard = crc
        currentDisplayedMonth = month

        crcName.text = crc.name
        crcOperator.text = crc.operator.name
        crcOperatorIcon.image = Image(Files.CRC_OPERATOR_ICONS_PATH + crc.operator.icon)

        val limit = crc.maxDebt
        val pendingPayments = creditCardService.getTotalPendingPaymentsByCreditCard(crc.id!!)
        val limitAvailable = creditCardService.getAvailableCredit(crc.id!!)
        val rebate = crc.availableRebate

        limitLabel.text = UIUtils.formatCurrency(limit)
        availableRebateLabel.text = UIUtils.formatCurrency(rebate)

        availableRebateLabel.style =
            when {
                rebate > BigDecimal.ZERO -> GREEN_LABEL_STYLE
                else -> BLACK_LABEL_STYLE
            }

        pendingPaymentsLabel.text = UIUtils.formatCurrency(pendingPayments)
        availableLimitLabel.text = UIUtils.formatCurrency(limitAvailable)

        val limitProgress =
            when {
                limit.isZero() -> BigDecimal.ZERO
                else -> pendingPayments.divide(limit, 2, RoundingMode.HALF_UP)
            }

        limitProgressBar.progress = limitProgress.toDouble()
        limitProgressLabel.text = UIUtils.formatPercentage(limitProgress.multiply(BigDecimal(100)))

        dueDateLabel.text = crc.billingDueDay.toString()
        closureDayLabel.text = crc.closingDay.toString()
        nextInvoiceLabel.text = UIUtils.formatShortMonthYear(creditCardService.getNextInvoiceDate(crc))

        updateInvoiceInfo()

        return rootVBox
    }

    fun updateInvoiceInfo() {
        invoiceMonthNavigatorBarLabel.text = UIUtils.formatShortMonthYear(currentDisplayedMonth)

        val totalDebts =
            creditCardService.getInvoiceAmount(creditCard.id!!, currentDisplayedMonth)

        invoiceMonthLabel.text = UIUtils.formatCurrency(totalDebts)

        invoiceStatusLabel.text =
            UIUtils.translateCreditCardInvoiceStatus(
                creditCardService.getInvoiceStatus(creditCard.id!!, currentDisplayedMonth),
            )
    }
}
